src/shared/logger.ts

```diff
@@ -5,139 +5,207 @@
 
 'use strict'
 
+import * as os from 'os'
 import * as path from 'path'
-import { MESSAGE } from 'triple-beam'
 import * as vscode from 'vscode'
 import * as nls from 'vscode-nls'
 import * as winston from 'winston'
 import * as Transport from 'winston-transport'
-import { SettingsConfiguration } from './settingsConfiguration'
+import { extensionSettingsPrefix } from './constants'
+import { DefaultSettingsConfiguration, SettingsConfiguration } from './settingsConfiguration'
 
-// TODO: Add ability to handle arbitrary values, through meta parameter or LogEntry parameter
 const localize = nls.loadMessageBundle()
 
-const logFormat = winston.format.printf(({ level, message, timestamp }) => {
-    return `${timestamp} [${level.toUpperCase()}]: ${message}`
-})
-
-const errFormat = winston.format( info => {
-    if (info.meta && info.meta instanceof Error) {
-        info.message = `${info.message} ${info.meta.stack}`
-    }
+const LOG_RELATIVE_PATH: string = path.join('Code', 'logs', 'aws_toolkit')
+const DEFAULT_LOG_LEVEL: string = 'info'
+const DEFAULT_LOG_NAME: string = `aws_toolkit_${makeDateString()}.log`
+const DEFAULT_OUTPUT_CHANNEL: vscode.OutputChannel = vscode.window.createOutputChannel('AWS Toolkit Logs')
+
+let defaultLogger: Logger
+
+export interface Logger {
+    logPath?: string
+    outputChannel?: vscode.OutputChannel
+    level: string
+    debug(...message: ErrorOrString[]): void
+    verbose(...message: ErrorOrString[]): void
+    info(...message: ErrorOrString[]): void
+    warn(...message: ErrorOrString[]): void
+    error(...message: ErrorOrString[]): void
+    releaseLogger(): void
+}
 
-    return info
-})
+export interface LoggerParams {
+    outputChannel?: vscode.OutputChannel
+    logPath?: string
+    logLevel?: string
+}
 
-let logSettings: LogSettings
+// creates default logger object
+// can overwrite defaultLogger if you include params
+export function initialize(params?: LoggerParams): Logger {
```

Comment:
I think this is fine as long as each test only calls initialize once, because our mocha tests are executed in serial.

If not, we may need to wrap this and `getLogger` with `async-lock`.

Reply:
Does that only work with async functions, or does it force a block of code to run synchronously? There's no actually-written-as-async functionality built into this code (the exception is the registered VS Code open-in-editor function, but that's completely separate from the core logging functionality). There's a quasi-async required if you want to check a file's output, but there's no way to confirm this creation through Winston's functionality, either through async or callback. My workaround for this was the `waitForLogFile` function in `logger.test.ts`, but this is only required if you need to check the logs output--Winston doesn't care about whether or not the file is created to function.

If you see this as a significant concern, I can move `waitForLogFile` to `loggerUtils.ts` so we can use this functionality elsewhere.

Reply:
No action required here, I was just thinking out-loud 😛